tsl/src/continuous_aggs/common.c

```diff
@@ -0,0 +1,1379 @@
+						custom_origin = true;
+						if (IsA(lfourth(fe->args), Const))
+						{
+							tbinfo->origin =
+								DatumGetTimestampTz(castNode(Const, lfourth(fe->args))->constvalue);
+						}
+						/* could happen in a statement like time_bucket('1h', .., 'utc', origin =>
+						 * ...) */
+						else if (IsA(lfourth(fe->args), NamedArgExpr))
+						{
+							Const *constval =
+								check_time_bucket_argument(lfourth(fe->args), "fourth");
+
+							tbinfo->origin = DatumGetTimestampTz(constval->constvalue);
+						}
+					}
+			}
+			if (custom_origin && TIMESTAMP_NOT_FINITE(tbinfo->origin))
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("invalid origin value: infinity")));
+			}
+
+			/*
+			 * We constify width expression here so any immutable expression will be allowed.
+			 * Otherwise it would make it harder to create caggs for hypertables with e.g. int8
+			 * partitioning column as int constants default to int4 and so expression would
+			 * have a cast and not be a Const.
+			 */
+			width_arg = linitial(fe->args);
+
+			if (IsA(width_arg, NamedArgExpr))
+				width_arg = (Node *) castNode(NamedArgExpr, width_arg)->arg;
+
+			width_arg = eval_const_expressions(NULL, width_arg);
+			if (IsA(width_arg, Const))
+			{
+				Const *width = castNode(Const, width_arg);
+				tbinfo->bucket_width_type = width->consttype;
+
+				if (width->consttype == INTERVALOID)
+				{
+					tbinfo->interval = DatumGetIntervalP(width->constvalue);
+					if (tbinfo->interval->month != 0)
+						tbinfo->bucket_width = BUCKET_WIDTH_VARIABLE;
+				}
+
+				if (tbinfo->bucket_width != BUCKET_WIDTH_VARIABLE)
+				{
+					/* The bucket size is fixed. */
+					tbinfo->bucket_width =
+						ts_interval_value_to_internal(width->constvalue, width->consttype);
+				}
+			}
+			else
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("only immutable expressions allowed in time bucket function"),
+						 errhint("Use an immutable expression as first argument to the time bucket "
+								 "function.")));
+
+			if (tbinfo->interval && tbinfo->interval->month)
+			{
+				tbinfo->bucket_width = BUCKET_WIDTH_VARIABLE;
+			}
+		}
+	}
+
+	if (tbinfo->bucket_width == BUCKET_WIDTH_VARIABLE)
+	{
+		/* Variable-sized buckets can be used only with intervals. */
+		Assert(tbinfo->interval != NULL);
+
+		if ((tbinfo->interval->month != 0) &&
+			((tbinfo->interval->day != 0) || (tbinfo->interval->time != 0)))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("invalid interval specified"),
+					 errhint("Use either months or days and hours, but not months, days and hours "
+							 "together")));
+		}
+	}
+
+	if (!found)
+		elog(ERROR, "continuous aggregate view must include a valid time bucket function");
+}
+
+static bool
+cagg_agg_validate(Node *node, void *context)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, Aggref))
+	{
+		Aggref *agg = (Aggref *) node;
```

Comment:
Nit: my pet peeve is that you do not need an explicit cast for pointers from and to `void*`. Feel free to ignore.

```suggestion
		Aggref *agg = node;
```